### ISSUE TITLE
model.backward在iter == arg.eval-interval时卡住无法继续

### ISSUE THREAD

**Issue by xikaluo** [THUDM/GLM#17]
您好，我在尝试运行GLM finetune代码时遇到了这样一个问题：我在其它iteration时都可以正常进行forward, backward和optimize step，但是当iteration index == arg.eval-iterval时，代码会在train_utils.py里的backward_step方法中的model.backward(loss)步骤卡住。希望您能帮我解答这个问题，谢谢～

deepspeed参数：
{
    "train_micro_batch_size_per_gpu": 4,
    "gradient_accumulation_steps": 1,
    "steps_per_print": 50,
    "gradient_clipping": 1.0,
    "zero_optimization": {
      "stage": 2,
      "contiguous_gradients": false,
      "overlap_comm": true,
      "reduce_scatter": true,
      "reduce_bucket_size": 5e7,
      "allgather_bucket_size": 5e7,
      "cpu_offload": true
    },
    "zero_allow_untested_optimizer": true,
    "fp16": {
      "enabled": true,
      "loss_scale": 0,
      "loss_scale_window": 1000,
      "hysteresis": 2,
      "min_loss_scale": 1
    },
    "optimizer": {
      "type": "Adam",
      "params": {
        "lr": 5e-6,
        "betas": [
          0.9,
          0.95
        ],
        "eps": 1e-8,
        "weight_decay": 1e-2
      }
    },
    "activation_checkpointing": {
      "partition_activations": false,
      "contiguous_memory_optimization": false
    },
    "wall_clock_breakdown": false
}

model config：
使用GLM-10B-zh，参数为
MODEL_TYPE="blocklm-10B"
MODEL_ARGS="--block-lm \
            --cloze-eval \
            --task-mask \
            --num-layers 48 \
            --hidden-size 4096 \
            --num-attention-heads 64 \
            --max-position-embeddings 1024 \
            --tokenizer-type ChineseSPTokenizer \
            --load-pretrained ${CHECKPOINT_PATH}"

任务参数：
TRAIN_ARGS="--epochs 1 \
            --batch-size 64 \
            --lr 1e-5 \
            --lr-decay-style linear \
            --warmup 0.06 \
            --weight-decay 1.0e-1 \
            --label-smoothing 0.1"

COMMON_ARGS="--save-interval 10000 \
             --log-interval 50 \
             --eval-interval 5 \
             --eval-iters 100 \
             --eval-epoch 2"

TASK_ARGS="--src-seq-length 608 \
           --tgt-seq-length 15 \
           --min-tgt-length 4 \
           --length-penalty 0.7 \
           --no-repeat-ngram-size 3 \
           --num-beams 2 \
           --select-topk \
           --eval-batch-size 1"

gpu：16G DCU卡

其中，每次代码在运行到iteration_ == 5，也就是第6个iter的时候会卡住

**Comment by duzx16** [comment]
The program takes a synchronization for all the processes at `model.backward`. If it is stuck here, maybe some processes haven't reached the line.

model.backward需要所有的进程同步，卡住说明有的进程没有执行到这一步，可以从这个方向去检查。

**Comment by JiaquanYe** [comment]
meet same issue
my question is model.backward(loss) work fine in when iteration_ % eval_interval !=0, but stuck on iteration_ % eval_interval ==0 (GPU run 100%).
it seems to be related to eval_interval  setting (but it didn't run validation dataset at all, cause valid_dataloader is None)

**Comment by JiaquanYe** [comment]
我运行的是ds_finetune_seq2seq的例子，通过调试发现，原因可能是，在Model_Parallel模式运行时，mp_rank==0的进程与mp_rank!=0的进程，其valid_dataloader分配不一致造成的，如下是源码的截取，可见mp_rank==0，通过train_valid_datasets_provider构建了一个valid_dataloader为None的实例，而mp_rank!=0的进程，则通过FakeDataloader构建了一个非None的实例
<img width="1013" alt="MP_issue_hang1" src="https://user-images.githubusercontent.com/18445710/227754856-73465d1f-a7c2-44e2-8ce1-1402a4d90142.png">

所以当在iter == arg.eval-interval时，mp_rank==0的进程，args.eval_interval and valid_dataloader is not None and args.iteration % args.eval_interval == 0 其判断为False，而mp_rank!=0的进程为True（valid_dataloader 不同导致的），这样就导致在model.backward(loss)时无法同步，见下图
<img width="1094" alt="MP_issue_hang" src="https://user-images.githubusercontent.com/18445710/227755110-e621b1a6-69bc-4c4f-ad63-4c5e2a61f4d9.png">


也请问一下是出于什么考虑需要把mp_rank!=0与mp_rank==0构建不同的valid_dataloader呢？
